aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move

```diff
@@ -0,0 +1,661 @@
+    ///   (2) signature share (for a multisignature or aggregate signature)
+    struct Signature has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    /// An aggregation of BLS signatures. This can be either a:
+    ///   (4) aggregated signature (i.e., an aggregation of signatures s_i, each on a message m_i)
+    ///   (3) multisignature (i.e., an aggregation of signatures s_i, each on the same message m)
+    ///
+    /// We distinguish between a Signature type and a AggrOrMultiSignature type to prevent developers from interchangeably
+    /// calling `verify_multisignature` and `verify_signature_share` to verify both multisignatures and signature shares,
+    /// which could create problems down the line.
+    struct AggrOrMultiSignature has copy, drop, store {
+        bytes: vector<u8>
+    }
+
+    /// Creates a new public key from a sequence of bytes.
+    public fun public_key_from_bytes(bytes: vector<u8>): Option<PublicKey> {
+        if (validate_pubkey_internal(bytes)) {
+            option::some(PublicKey {
+                bytes
+            })
+        } else {
+            option::none<PublicKey>()
+        }
+    }
+
+    /// Creates a new proof-of-possession (PoP) which can be later used to create a PublicKeyWithPoP struct,
+    public fun proof_of_possession_from_bytes(bytes: vector<u8>): ProofOfPossession {
+        ProofOfPossession {
+            bytes
+        }
+    }
+
+    /// Creates a PoP'd public key from a normal public key and a corresponding proof-of-possession.
+    public fun public_key_from_bytes_with_pop(pk_bytes: vector<u8>, pop: &ProofOfPossession): Option<PublicKeyWithPoP> {
+        if (verify_proof_of_possession_internal(pk_bytes, pop.bytes)) {
+            option::some(PublicKeyWithPoP {
+                bytes: pk_bytes
+            })
+        } else {
+            option::none<PublicKeyWithPoP>()
+        }
+    }
+
+    /// Creates a new signature key from a sequence of bytes. Does not check the signature for prime-order subgroup
+    /// membership since that is done implicitly during verification.
+    public fun signature_from_bytes(bytes: vector<u8>): Signature {
+        Signature {
+            bytes
+        }
+    }
+
+    /// Given a vector of public keys with verified PoPs, combines them into an *aggregated* public key which can be used
+    /// to verify multisignatures using `verify_multisignature` and aggregate signatures using `verify_aggregate_signature`.
+    /// Returns 'None' if no public keys are given as input.
+    /// Does not abort.
+    public fun aggregate_pubkeys(public_keys: vector<PublicKeyWithPoP>): Option<AggrPublicKeysWithPoP> {
+        let (bytes, success) = aggregate_pubkeys_internal(public_keys);
+        if (success) {
+            option::some(
+                AggrPublicKeysWithPoP {
+                    bytes
+                }
+            )
+        } else {
+            option::none<AggrPublicKeysWithPoP>()
+        }
+    }
```

Comment:
why wouldn't this abort?

Reply:
...because I don't understand the semantics of abort / how bad they are / how gas metering works during aborts.